### ISSUE TITLE
Usage with near-operation-file-preset?

### ISSUE THREAD

**Issue by leggomuhgreggo** [zhouzi/graphql-codegen-factories#17]
Hello! I was wondering if there's any way to use the plugin with [`near-operation-file-preset` plugin](https://www.graphql-code-generator.com/plugins/near-operation-file-preset), so that the factories are shaped/co-located with respect to individual `.graphql` documents.

Thanks!

**Comment by zhouzi** [comment]
Hey @leggomuhgreggo 👋 I never used `near-operation-file-preset` but it looks pretty cool and it shouldn't be too hard to support for it. Just to be sure, you'd like to create factories for operations and have them colocated with the source files?

So the following:

```graphql
mutation CreateUser($name: String!) {
  createUser(name: $name) {
    id
    name
  }
}
```

Should lead to:

```typescript
function createCreateUserMock(props: Partial<CreateUserMutation>): CreateUserMutation {
  return {
    ...props,
    __typename: 'Mutation',
    createUser: {
      __typename: 'User',
      id: '',
      name: ''
    }
  };
}
```

Is that correct or do you have a different use case?

**Comment by leggomuhgreggo** [comment]
Golly, thanks for being so generous!

Yes, that's the use case. That way the tests for a given query-connected component, can use the factories with strictly typed component interfaces. I was getting TS errors with the full-type factories.

Let me know if there's anything I can do to help! Thanks for the cool library!

**Comment by zhouzi** [comment]
I just merged #18, it adds support for split output which is required to fix this issue. The next step now is to implement support for operations. Hopefully it doesn't mean reimplementing the [`typescript-operations`](https://github.com/dotansimha/graphql-code-generator/tree/master/packages/plugins/typescript/operations) plugin, we will see.

**Comment by leggomuhgreggo** [comment]
That makes sense -- interesting to see the diff

(and I salute your attention to documentation!)

**Comment by zhouzi** [comment]
I spent some time looking into adding support for operations and it's more complex than I imagined. I will probably pause the development of this feature so I thought I would write down my notes here. In case someone wants to take over or for my future self.

<details>

<summary>Notes</summary>

The challenge here is that there is a lot of variety in how you can write an operation. For example, you can use fragments:

```graphql
mutation CreateBlogPost($title: String!, $content: String!) {
  createBlogPost(title: $title, content: $content) {
    ...BlogPostFragment
  }
}
fragment BlogPostFragment on BlogPost {
  title
  content
  author {
    ...AuthorFragment
  }
}
fragment AuthorFragment on Author {
  name
  followers {
    ...FollowerFragment
  }
}
fragment FollowerFragment on Follower {
  name
}
```

In order to generate a `createCreateBlogPostMutationMock` factory we first need to resolve all the fragments and lookup their types in the schema. `graphql-codegen` does offer some utilities to help with that but they are mostly targeted at generating types so they are hardly compatible with what this plugin wants to generate.

If we were to go further, here is what is required:

1. The package must expose two plugins: one for the schema and another one for operations. Just like there is a `typescript` and a `typescript-operations` plugin. That's because the former visits the schema and the latter visits documents. If they are merged together, it would generate the factories for the schema as many times as there are documents.
2. Ideally the operations visitor would be based off [`BaseDocumentsVisitor`](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/other/visitor-plugin-common/src/base-documents-visitor.ts) but it [generates types](https://github.com/dotansimha/graphql-code-generator/blob/4d2b101a546f4b106d631d64afea68a1918ec889/packages/plugins/other/visitor-plugin-common/src/base-documents-visitor.ts#L239-L260) so we have to go with the lower level [`BaseVisitor`](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/other/visitor-plugin-common/src/base-visitor.ts). We probably have to duplicate some code from `BaseDocumentsVisitor` which integrates helpful utilities.
3. We should be able to leverage [`SelectionSetObject`](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/other/visitor-plugin-common/src/selection-set-to-object.ts) to deal with selection sets (e.g to resolve fragments). It seems to be targeted at generating types as well though, so we would need to give it a try. If it is not compatible, that's a deal breaker because this class does a lot and having to duplicate it would make the plugin super fragile.

With that being said, preferably we would contribute to `graphql-codegen` and see if we can make their utilities less types-oriented.

</details>

In the mean time, users can still create their own factories based on the ones generated from the schema. If we take back the example operation from above (see notes), here's what it could look like:

```typescript
import { createBlogPostMock } from "./generated/factories";

function createCreateBlogPostMutationMock() {
  return {
    createBlogPost: createBlogPostMock({})
  };
}
```

It's not perfect but seems to be a decent workaround.

**Comment by ertrzyiks** [comment]
Since I'm also investigating this topic let me share my findings.

[BaseDocumentsVisitor](https://github.com/dotansimha/graphql-code-generator/blob/master/packages/plugins/other/visitor-plugin-common/src/base-documents-visitor.ts) is the best visitor to start with. Although it's indeed targeted at types, the implementation covers all the needs that we have to generate factories. 

I found a way to inject into the generation process to replace the currently serializes the operation into a type, [wrote it down as an answer to this question](https://github.com/dotansimha/graphql-code-generator/discussions/7646).

We can create a subclass of `SelectionSetToObject` and make `transformSelectionSet` use the operation data to generate the factory, not a type. The original `SelectionSetToObject` implementation includes all the helpers to resolve the selection complexity regarding fragments.

```
class SelectionSetToFactory<Config extends ParsedDocumentsConfig = ParsedDocumentsConfig> extends SelectionSetToObject<Config> {
  public transformSelectionSet() {

    return 'function (){ }'
  }

  createNext(parentSchemaType: GraphQLNamedType, selectionSet: SelectionSetNode) {
    return new SelectionSetToFactory(this._processor, this._scalars, this._schema, this._convertName.bind(this), this._getFragmentSuffix.bind(this), this._loadedFragments, this._config, parentSchemaType, selectionSet);
  }
}
```

**Comment by zhouzi** [comment]
Thanks for sharing @ertrzyiks 🙏 I went ahead and dug deeper with your insights. I was able to generate something, it's not quite right yet but I am getting there. I had to tweak a few things though.

I chose to extend `BaseVisitor` instead of `BaseDocumentsVisitor` because with the latter, we need to override both `FragmentDefinition` and `OperationDefinition` methods. The only useful thing left is the `handleAnonymousOperation` method but it's private so we can't use it in an extended class (it should probably be a protected method instead). Also, this visitor expects a ton of configuration that probably does not make sense for our use case.

Then as you suggested, I had to create a subclass of `SelectionSetToObject`. I also had to create a subclass of `BaseSelectionSetProcessor` which is `SelectionSetToObject`'s first parameter. I am now trying to make sense of all the configuration that is expected by `SelectionSetToObject` but it seems related to types for the most part. So we are probably fine providing default values.

I will open a PR soon 👌

**Comment by zhouzi** [comment]
I just published [`graphql-codegen-factories@1.0.0-beta.1`](https://github.com/zhouzi/graphql-codegen-factories/releases/tag/1.0.0-beta.1) with support for operations and the `near-operation-file-preset` plugin. Feel free to give it a try and open an issue if you find any bug 🙏
